### PR TITLE
fetchpatch2: init

### PR DESCRIPTION
**Pull request by Artturin**
allows us to use the new features of patchutils without having to reset
all fetchpatch hashes in nixpkgs

https://github.com/NixOS/nixpkgs/issues/32084

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
In principle I like the approach.  Before merging, we perhaps should think if there's some other potentially hash-changing... change that we might want to have as well.

File: pkgs/test/default.nix [hunk 1]
```diff
@@ -31,6 +31,7 @@ with pkgs;
 
   fetchurl = callPackages ../build-support/fetchurl/tests.nix { };
   fetchpatch = callPackages ../build-support/fetchpatch/tests.nix { };
+  fetchpatch2 = callPackages ../build-support/fetchpatch/tests.nix { fetchpatch = fetchpatch2; };
```
Comment (SuperSandro2000) [inline pkgs/test/default.nix:34]: I think we should move this right into the test then we could also reduce the changes in all-packages.

**Comment by roberth** [comment]
> In principle I like the approach. Before merging, we perhaps should think if there's some other potentially hash-changing... change that we might want to have as well.

https://github.com/NixOS/nixpkgs/issues/48567 is the only issue I've found, but can be largely resolved by hashing parameters into a `name` prefix or suffix. That's not a _content_ hash changing operation though, so there's no benefit in bundling that one.

**Review by roberth — APPROVED** [review]
LGTM, but needs more eyes.

**Review by IvarWithoutBones — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -691,10 +691,20 @@ with pkgs;
...
-  fetchpatch = callPackage ../build-support/fetchpatch { }
-    // {
-      tests = pkgs.tests.fetchpatch;
-    };
+  fetchpatch = callPackage ../build-support/fetchpatch {
+    # 0.3.4 would change hashes: https://github.com/NixOS/nixpkgs/issues/25154
+    patchutils = buildPackages.patchutils_0_3_3;
+  } // {
+    tests = pkgs.tests.fetchpatch;
+    version = 1;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:699]: Since fetchurl now support pname+version we could set this via that but not super important.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (roberth) [reply pkgs/top-level/all-packages.nix:699]: That'd be on the result of fetchpatch, whereas the test needs it to have a version for the function itself.
I think having the version on the result (derivation) would be confusing. (Why does my patch have a version I didn't specify?)

File: pkgs/test/default.nix (hunk 1 above)
Comment (roberth) [reply pkgs/test/default.nix:34]: I think the current solution is decent. If we need something else, we can change it.

**Comment by Artturin** [comment]
perhaps we should do the the downloading in a fod and then the patchutils stuff in a non fod to prevent hash incompabilities when we update patchutils

**Comment by jtojnar** [comment]
The main point of `fetchpatch` is making FOD for generated patches that can change.

**Pull request merged**